wordpress-mobile/WordPress-Android
# [Jetpack Content Migration Flow] Add tracking events

Fixes #17572
Fixes #17574

This PR adds tracking for the Migration Flow in the Jetpack App, and unifies tracking via `ContentMigrationAnalyticsTracker`.

It tackles both issue linked above.

To test:

#### 1️⃣ Test migration happy flow tracking events

1. Build & install the WordPress app `wordpressWasabiDebug` flavor, open app and login
2. Install the Jetpack app of the same flavor (e.g. `jetpackWasabiDebug`)
3. Open the Jetpack app
4. The Welcome Screen of the Jetpack migration flow should be shown
   **Verify** in logcat that the following tracking events are logged:
   1. `jpmigration_wordpressapp_detected`
   ```
   🔵 Tracked: jpmigration_wordpressapp_detected, Properties: {"compatible":true}
   ```
   2. `jpmigration_welcome_screen_shown`
   ```
   🔵 Tracked: jpmigration_welcome_screen_shown
   ``` 
5. Tap `Need Help`
   - **Verify** the `jpmigration_welcome_screen_help_button_tapped` tracking event is logged:
   ```
   🔵 Tracked: jpmigration_welcome_screen_help_button_tapped
   ```
6. Go back
7. Tap on the user avatar image in the top right corner
   **Verify** the `jpmigration_welcome_screen_avatar_tapped` is tracked
8. Go back
9. Tap the `Continue` button on the Welcome Screen
10. The Notifications screen of the flow should be shown
    **Verify** the following events are tracked:
    1. `jpmigration_welcome_screen_continue_button_tapped`
    2. `jpmigration_notifications_screen_shown`
11. Tap the `Continue` button on the Notifications screen
12. The success/thanks screen of the flow should be shown
    **Verify** the following events are tracked:
    1. `jpmigration_notifications_screen_continue_button_tapped`
    2. `jpmigration_thanks_screen_shown`
13. Tap the `Finish` button on the success/thanks screen
14. The `My Site` screen should be shown
    **Verify** the `jpmigration_thanks_screen_finish_button_tapped` is tracked
15. On the `My Site` screen tap the `Please delete the WordPress app` card
16. The "You no longer need the WordPress app" screen should be shown
    **Verify** the following events are tracked:
    1. `jpmigration_please_delete_wordpress_card_tapped`
    2. `jpmigration_please_delete_wordpress_screen_shown`
17. Tap `Need Help` on the screen
    **Verify** the `jpmigration_please_delete_wordpress_help_button_tapped` event is tracked
18. Go back
19. Tap `Got it` on the screen to delete WordPress app
20. The `My Site` screen should be shown
    **Verify** the `jpmigration_please_delete_wordpress_gotit_tapped` event is tracked

<details>
<summary><h4>2️⃣ Test incompatible WordPress app detected</h4></summary>

1. Prerequisite: WordPress app should be installed and logged in
2. Uninstall Jetpack App if installed
3. In `JetpackAppMigrationFlowUtils` change `minimumSupportedVersion` to `21.4`
   > This is a hack to mimic real-life scenarios where the Jetpack app requires a higher version of the WordPress than what's currently installed.
4. Build and install the Jetpack app of the same flavor that the installed WordPress app is (e.g. `wasabiDebug`)
5. Open the Jetpack App
6. The migration flow should **not** be shown, instead the Landing screen should appear
   **Verify** the `jpmigration_wordpressapp_detected` event is tracked with the "compatible" property false
   ```
   🔵 Tracked: jpmigration_wordpressapp_detected, Properties: {"compatible":false}
   ```
</details>

<details>
<summary><h4>3️⃣ Test error screen events tracking in the migration flow</h4></summary>

1. Prerequisite: WordPress app should be installed and logged in
2. Uninstall Jetpack App if installed
3. In `build.gradle` change `JETPACK_LOCAL_USER_FLAGS` build config value to `false`:
   ```
   buildConfigField "boolean", "JETPACK_LOCAL_USER_FLAGS", "false"
   ```
   **Note** _This is a hardcode solution to make the migration flow always fail._
4. Sync project
5. Build and install the Jetpack app of the same flavor that the installed WordPress app is (e.g. `wasabiDebug`)
6. Open the Jetpack app
7. The Error screen of the migration flow should be shown
   **Verify** the `jpmigration_error_screen_shown` event is tracked
8. Tap the `Need Help` button
   **Verify** the `jpmigration_error_screen_help_button_tapped` is tracked
9. Go back
10. Tap the `Try Again` button
   **Verify** the `jpmigration_error_screen_retry_button_tapped` is tracked

</details>

## Regression Notes

1. Potential unintended areas of impact
   Jetpack content migration flow.

12. What I did to test those areas of impact (or what existing automated tests I relied on)
   Manual testing, & relied on current automated tests.

13. What automated tests I added (or what prevented me from doing so)
   Added tests for tracking in:
   - `JetpackAppMigrationFlowUtilsTest`
   - `MySiteViewModelTest`
   - `JetpackMigrationViewModelTest`

PR submission checklist:
